xds/internal/balancer/clusterresolver/e2e_test/balancer_test.go

```diff
@@ -302,3 +313,217 @@ func (s) TestErrorFromParentLB_ResourceNotFound(t *testing.T) {
+		priorityBuilder: b,
+	}
+}
+
+func (b *testChildPolicyBalancerBuilder) Build(cc balancer.ClientConn, opts balancer.BuildOptions) balancer.Balancer {
+	priorityLB := b.priorityBuilder.Build(cc, opts)
+	return &testChildPolicyBalancer{
+		Balancer:  priorityLB,
+		scStateCh: b.scStateCh,
+	}
+}
+
+func (b *testChildPolicyBalancerBuilder) Name() string {
+	return priority.Name
+}
+
+func (b *testChildPolicyBalancerBuilder) ParseConfig(lbCfg json.RawMessage) (serviceconfig.LoadBalancingConfig, error) {
+	select {
+	case b.cfgCh <- lbCfg:
+	default:
+	}
+	return b.priorityBuilder.(balancer.ConfigParser).ParseConfig(lbCfg)
+}
+
+type testChildPolicyBalancer struct {
+	balancer.Balancer
+	scStateCh chan balancer.SubConnState
+}
+
+func (b *testChildPolicyBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	select {
+	case b.scStateCh <- state:
+	default:
+	}
+
+	b.Balancer.UpdateSubConnState(sc, state)
+}
+
+func (s) TestSubConnStateChangePropagationToChildPolicy(t *testing.T) {
+	// Unregister the priority balancer builder for the duration of this test,
+	// and register a policy under the same name which makes subConn state
+	// changes pushed to it available to the test.
+	priorityBuilder := balancer.Get(priority.Name)
+	internal.BalancerUnregister(priorityBuilder.Name())
+	testChildPolicy := newTestChildPolicyBalancerBuilder(priorityBuilder)
+	balancer.Register(testChildPolicy)
+	defer balancer.Register(priorityBuilder)
+
+	managementServer, nodeID, bootstrapContents, _, cleanup := e2e.SetupManagementServer(t, e2e.ManagementServerOptions{})
+	defer cleanup()
+
+	server := stubserver.StartTestService(t, nil)
+	defer server.Stop()
+
+	// Configure cluster and endpoints resources in the management server.
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Clusters:       []*v3clusterpb.Cluster{e2e.DefaultCluster(clusterName, edsServiceName, e2e.SecurityLevelNone)},
+		Endpoints:      []*v3endpointpb.ClusterLoadAssignment{e2e.DefaultEndpoint(edsServiceName, "localhost", []uint32{testutils.ParsePort(t, server.Address)})},
+		SkipValidation: true,
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if err := managementServer.Update(ctx, resources); err != nil {
+		t.Fatal(err)
+	}
+
+	// Create xDS client, configure cds_experimental LB policy with a manual
+	// resolver, and dial the test backends.
+	cc, cleanup := setupAndDial(t, bootstrapContents)
+	defer cleanup()
+
+	// Spawn a goroutine which ensures that subConn state changes are propagated
+	// to the child policy. We *only* wait for READY state since we expect to
+	// get to that state when the below RPC succeeds.
+	doneCh := make(chan error, 1)
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				doneCh <- fmt.Errorf("test timeout expired when waiting for child policy to see a READY subConn")
+				return
+			case state := <-testChildPolicy.scStateCh:
```

Comment:
Is this not racy? I.e. after spawning goroutine for this check, we're not guaranteed the processory actually hits this select, and the other operations from lines 421 - 428 can execute without this select running in a separate goroutine, and the line 362 hits, and we lose the Update?

Reply:
Great find. Switched to using a `buffer.Unbounded` because I want two properties here:
- no state updates should be lost (as you pointed out)
- the `UpdateSubConnState` method should not block when writing on the channel because the test goroutine isn't reading anymore. This can be lead to test failures as well with the `leakchecker`.

I didn't switch the other channel to a `buffer.Unbounded` because we don't have that issue there (at least with the current usage).

Reply:
Right, I was actually going to suggest this pattern. In order to not lose state, we need an arbitrary length buffer, and keep reading until READY received. So this change lgtm.